toolchain/install/install_paths.cpp

```diff
@@ -0,0 +1,108 @@
+  // Map from the executable path from the executable path to an install prefix
+  // path.
+
+  // TODO: Detect a Windows executable path and use custom logic to map to the
+  // correct install prefix for that platform.
+
+  // First, see if the binary is in a `bin` directory. If so, remove it
+  // assuming this is a FHS-like install prefix. We start with this pattern both
+  // because we expect it to be the "production" pattern and so useful to
+  // involve a minimal number of filesystem accesses, and because even in
+  // testing or development builds, when we _can_ use this pattern, we prefer it
+  // to match the behavior of genuine installations as closely as possible.
+  prefix_.assign(exe_path);
+  llvm::sys::path::remove_filename(prefix_);
+  if (auto rbegin_it = llvm::sys::path::rbegin(prefix_),
+      rend_it = llvm::sys::path::rend(prefix_);
+      rbegin_it != rend_it) {
+    // Step over a trailing `.` signifying the current directory. Whether this
+    // occurs for paths is ambiguous between the documentation and the observed
+    // behavior, so this code works to be resilient.
+    if (*rbegin_it == ".") {
+      ++rbegin_it;
+    }
+    if (rbegin_it != rend_it && *rbegin_it == "bin") {
+      // Rather than trying to remove the `bin` directory, walk upwards by
+      // appending `..`. We also assume POSIX style at this point.
+      llvm::sys::path::append(prefix_, llvm::sys::path::Style::posix, "../");
+      return;
+    }
+  }
+
+  // Next check for a Bazel runfiles directory relative to the binary name. This
+  // is used for `genrule` style execution within Bazel and for direct execution
+  // of tests or other just-built binaries.
+  prefix_.assign(exe_path);
+  prefix_.append(RunfilesSuffix);
+  if (llvm::sys::fs::is_directory(prefix_)) {
+    llvm::sys::path::append(prefix_, llvm::sys::path::Style::posix, Workspace,
+                            PrefixRoot);
+    CARBON_CHECK(llvm::sys::fs::is_directory(prefix_))
+        << "Found runfiles tree but not an install prefix directory!";
+    return;
+  }
+
+  // Lastly, check for a test runtimes tree environment variable. This will
+  // match when running inside a Bazel test's runfiles tree. We check this last
+  // of all in part because that allows testing of all of the prior patterns.
+  prefix_.assign(llvm::StringRef(getenv("TEST_SRCDIR")));
+  if (llvm::sys::fs::is_directory(prefix_)) {
+    llvm::sys::path::append(prefix_, llvm::sys::path::Style::posix, Workspace,
+                            PrefixRoot);
+    CARBON_CHECK(llvm::sys::fs::is_directory(prefix_))
+        << "Found runfiles tree but not an install prefix directory!";
+    return;
+  }
+
+  // Otherwise, we use whatever path remained from removing the filename of the
+  // executable path. This isn't a great fallback, but we don't have a lot of
+  // choices.
+  CARBON_VLOG()
+      << "Failed to detect a recognized install path, falling back to: "
+      << prefix_;
```

Comment:
I think this fallback might actually be dangerous and miss issues, particularly given the custom runfiles implementation. Can it verify that structure seems correct?

Reply:
Thanks, I see that. But it looks like `error()` is only checked in install_paths_test -- am I missing something? Note, `ErrorOr` does have some burden, but that's because it's forcing checks.

If you prefer to keep returning a finished `InstallPaths` when there's a structural error, you might consider something like `CheckMarkerFile() -> ErrorOr<Success>` instead of storing an `error()` state. Really there's not much need to store the error if it's not used, and it avoids an implication that the structure is acting on invalid state.

Reply:
FWIW, this seems fine to fix forward, and so based on approval going to merge as-is. It at least covers checking in all the tests. And it'll let me empty the stack of PRs. Happy to follow up and fix more cases, and make any further API tweaks here though if there are better APIs, especially if we have a better idea of how to handle diagnosing issues here in the driver itself (the only place where it seems a bit tricky).